### ISSUE TITLE
Feature request: configuration flag to drop invalid lookups

### ISSUE THREAD

**Issue by quasipedia** [y-ken/fluent-plugin-geoip#18]
One problem that I am experiencing with this otherwise flawless plugin, is that it will output `null` as the value for invalid IPs.  For example: looking up the geolocation of ``127.0.0.1`` will output `null` for both latitude and longitude.

The problem is that the data storage I am forwarding events to is unhappy with such values, as it expects valid coordinates for any record I am sending.

Unless there is some fluentd generic configuration setting I could use for this, I think that having a configuration flag such as for example:

    drop_invalid      true

would be very useful.  Such flag should simply omit to append the geoip info to the record if the lookup operation fails.

**Comment by y-ken** [comment]
Hi @quasipedia .
Thank you for suggestion.
How about using rewrite_tag_filter to drop records such as latitude is null(blank)?

```
# example of drop blank records
<match rewrite_tag.access>
  type rewrite_tag_filter
  rewriterule1 latitude ^$ clear
  rewriterule2 latitude .+ s3.access
</match>

<match s3.access>
  type s3
  ...snip...
</match>

<match clear>
  type null
</match>
``

cf. https://github.com/fluent/fluent-plugin-rewrite-tag-filter

**Comment by quasipedia** [comment]
Hi @y-ken! Thank you for the suggestion.  Indeed this was one of the options I explored (using that very plugin for other purposes), but I ended up finding a way to tell my storage layer to drop that specific field if both lat and lon are null.

Both solutions - though - are kind of workaround IMO:  mine is platform-dependent, and yours requires a whole new dependency and quite some write-up.

After having given some thought, I'd like to modify my initial suggestion though: **beside** having a `drop_invalid` flag, it would be useful to alternatively define a custom value for those records (one may wish for example to default the information to a specific location).  Something like having a series of values defined:

    city.default              Alghero
    latitude.default          34.2321
    longitude.default         11.3453
    country_code3.default     RUS
    country_code.default      IT
    country_name.default      "France"
    dma_code.default          <DROP>
    area_code.default         <DROP>
    region.default            null

The idea here would be that on failing to look-up any key you could either define a custom value (including `null`), or drop that part of the record entirely.

Anyhow: thank you for the prompt reply and your suggestion! Much appreciated! :)

**Comment by RolphH** [comment]
I'm facing the same issue and like to suggestion of adding either a drop_invalid value and/or the more specific solution of default values. If either one could be added that would save my day.

**Comment by y-ken** [comment]
Hi @RolphH , @quasipedia 

Thank you for your opinion.
I'm worry about the integrate auto schema product such as elasticsearch.
Using elasticsearch without mapping schema, It would be create schema with the first record fields.
To avoid such case, it would be better with many keys with null value.

But if it is the optional settings and disabled by default, it is worth serious consideration.
Which is the best beavior for this option?

## idea.1

won't add the geoip fields like `country_name`

* input

```
{"ip":"127.0.0.1"}
{"ip":"8.8.8.8"}
```

* output

```
{"ip":"127.0.0.1"}
{"ip":"8.8.8.8", "county_name":"France"}
```


## idea.2

drop whole line of record

* input

```
{"ip":"127.0.0.1"}
{"ip":"8.8.8.8"}
```

* output

```
{"ip":"8.8.8.8", "county_name":"France"}
```

Sincerely yours,
ken

**Comment by RolphH** [comment]
Hi @y-ken 

I'm currently using elasticsearch and the problem there is that is expects a geopoint (due to mapping). In the configuration I have something like:

> location_array  "[${longitude['clientip']},${latitude['clientip']}]"

With an IP like 127.0.0.1 this will result in 

> location_array "[null, null]"

I prefer to use idea.1 as I need the rest of the record too (basically idea.2 is what currently is happening as elasticsearch fails with an invalid geopoint)

**Comment by quasipedia** [comment]
@y-ken - Thanks for coming back to us.

**Idea number 1 is much better than idea 2**, IMO.  With idea 2 you actually lose valid data, while with idea 1 you simply don't generate erroneous one.

However a full-blown specification could foresee two things:

1. A series of default values: for each field that you can generate with the plugin, you can define a value that will be used for those IP whose lookup fails.
2. A setting (e.g.: `fallback_behaviour`) that you can set to two possible values: `ignore_record` (your "idea 2") or `skip_field` (your "idea 1"). This setting will **only be used for all those fields for which a default has _not_ been set, and for which the IP lookup fails**.

Also: **elasticsearch should _never_ be used without a proper template (mapping) in production environments**.  The fact elasticsearch has a dynamic schema generation is so only to make life easier for new developers approaching elasticsearch for the first time (exactly like the `zen` autodiscovery setting, which is also not to be used in production).  The reason is that ES has _dynamic schema_ but it is not _schemaless_: one the schema has been generated, each document _must_ conform to it.

Besides: ES is not able to auto-detect geohashes as such (it just mark them as strings), so a mapping/template is _needed_ to have ES understand certain geographical data formats.

In fact this is the problem that @RolphH (and I) are facing: we defined a geopoint (lat, lon as numbers) and when ES receives `null` it complains...

As a side note for @RolphH: the way to properly work around the problem you are experiencing, is to define a custom `null` value for the geopoint, using the `null_value` parameter in your mapping to specify that a `null` geo point is in fact a `(null, null)` tuple [I am out of my work VPN now, so I can't post a snippet of my mapping, sorry].

**Comment by RolphH** [comment]
Hi did you look into this request?

**Comment by y-ken** [comment]
I'm got it. I'm going to implement option with idea1 or considering changing the default behavior to return null or change nothing.

**Comment by Seraf** [comment]
Hello, I :+1: for this feature, elasticsearch produce a stacktrace for each entries containing a local ip in my nginx logs, it's a nightmare to read the elasticsearch log ...

**Comment by Seraf** [comment]
Hello,

@y-ken , any news about this feature please ?

**Comment by RolphH** [comment]
We solved it by adding a transform to the mapping as follows:

```
"transform": {
                    "script": "if (ctx._source['location_array'] == [null,null]) { ctx._source.remove('location_array'); }",
                    "lang": "groovy"
}
```

**Comment by y-ken** [comment]
Hi all

Good news.
Long we have kept you waiting, I've implement to solve this issue.
Please check it at the Pull-requests.
https://github.com/y-ken/fluent-plugin-geoip/pull/20

**Comment by y-ken** [comment]
Hi @quasipedia  @RolphH  @Seraf 

I have released fluent-plugin-geoip v0.5.0 now.
https://rubygems.org/gems/fluent-plugin-geoip

On td-agent2, you can install with following commands.

```
$ sudo /usr/sbin/td-agent-gem install fluent-plugin-geoip
```

And adding configuration `skip_adding_null_record on` for geoip match block.

```
<match foo.bar>
  @type geoip
  ...snip...
  skip_adding_null_record  true
```

**Comment by quasipedia** [comment]
I have move on to another role, but today a colleaugue is upgrading our machine to make use of this option.  Thanks a lot for having implemented it @y-ken !! :)
